esp32c6: PCNT Quadrature Encoder driver

Vajnar opened the pull request:
## Summary
This PR introduces Qradrature Encoder driver for ESP32C6 devkit C. The driver is based on `arch/xtensa/src/esp32/esp32_qencoder.c`. The main difference is the number of units which is 4 on ESP32C6 instead of 8 on Xtensa.

## Impact
It is possible to use `SENSORS_QENCODER` to interface with the driver. ~~This PR requires https://github.com/espressif/esp-hal-3rdparty/pull/4.~~

## Testing
The PR was tested on ESP32C6 devkit C with brushed DC motor with quadrature encoder outputs.

Co-authored-by: Pavel Pisa (<pisa@fel.cvut.cz>)

Vajnar commented:
cc: @ppisa

tmedicci commented:
@fdcavalcanti , please take a look too ;)

tmedicci reviewed (changes requested):
Hi @Vajnar ,

First of all, thank you for submitting it. It's very valuable.

I have to point some key points about it: whenever we provide a driver in `arch/risc-v/src/common/espressif/`, it's expected to be shared between all Espressif devices that implement that peripheral (including not-yet supported devices on NuttX). The important thing here is not to access any registers directly. Instead of using the HAL definitions from `components/soc/<chip>/include/soc/pcnt_reg.h`, we use the functions from `components/hal/<chip>/include/hal/pcnt_ll.h`: these interfaces are "well-behaviored" and can be used by the common driver across different devices. For inspiration on how this driver could be implemented, you can take a look at `components/driver/pcnt/pulse_cnt.c` of ESP-IDF.

We don't expect you to develop and test it for all devices (we can test on other devices for you and even suggest improvements), but this is only possible if we don't access directly, otherwise, the driver could not be used for other devices.

My recommendation: try to the HAL-provided functions (from HAL's components/hal/<chip>/include/hal/) and we can help you to test for other devices. If I'm not mistaken, only ESP32-H2 (considering Espressif's RISC-Vs) features PCNT peripheral.

fdcavalcanti commented on arch/risc-v/src/common/espressif/Kconfig:
```diff
@@ -486,6 +493,215 @@ config ESP_MCPWM
 		the Motor Control PWM peripheral.
 
+menu "Pulse Counter (PCNT) Configuration"
+	depends on ESP_PCNT
```
Pulse counter should only be available on certain Espressif chips (C3 for instance does not have PCNT peripheral)

Vajnar replied:
Could you, please, provide me on which chips it should depend?

fdcavalcanti replied:
PCNT support for available in: ESP32H2 and ESP32C6

Vajnar replied:
Thank you, I've added dependency.

fdcavalcanti commented on boards/risc-v/esp32c6/common/src/esp_board_qencoder.c:
```diff
@@ -0,0 +1,66 @@
+#include <arch/board/board.h>
+
+#include "espressif/esp_qencoder.h"
+//#include "chip.h"
```
Can remove this comment

Vajnar replied:
done

Vajnar commented:
Hi @tmedicci,

thank You for the valuable feedback. I removed all direct register accesses and replaced them with function calls from `components/hal/<chip>/include/hal/pcnt_ll.h`. Does it now look better?

The changes also make https://github.com/espressif/esp-hal-3rdparty/pull/4 not needed anymore, so I will close it.

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+ */
+
+#ifndef CONFIG_DEBUG_FEATURES
+#  undef CONFIG_DEBUG_SENSORS
```
DEBUG_SENSORS should not be active when DEBUG_FEATURES is undef. I think this is not needed.

Vajnar replied:
removed

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+
+/* Overall, RAM-based state structure */
+
+struct esp_lowerhalf_s
```
`struct qe_dev_lowerhalf_s` would be more appropriate. `esp_lowerhalf_s` is too generic

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+  const struct esp_qeconfig_s *config; /* static configuration */
+
+  bool inuse; /* True: The lower-half driver is in-use */
+
```
Can remove those empty lines

Vajnar replied:
done

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+/* Interrupt handling */
+
+#if 0 /* FIXME: To be implemented */
+static int esp_interrupt(int irq, void *context, void *arg);
```
What's wrong with this?

Vajnar replied:
The interrupt is not required for basic functionality, but could be interesting for future SW extension of pulse counter from 16 to 32 bits.

tmedicci commented:
> Hi @tmedicci,
> 
> thank You for the valuable feedback. I removed all direct register accesses and replaced them with function calls from `components/hal/<chip>/include/hal/pcnt_ll.h`. Does it now look better?
> 
> The changes also make [espressif/esp-hal-3rdparty#4](https://github.com/espressif/esp-hal-3rdparty/pull/4) not needed anymore, so I will close it.

That's fine, thanks @Vajnar !

I'm not sure how you tested it, please describe it in detail. Particularly, I would like to know which testing application you used (if public, of course) and how we can reproduce and validate the testing from our side.

In this same direction, can you please add a `defconfig` file for it in `nuttx/boards/risc-v/esp32c6/esp32c6-devkitc/configs/` and add an entry in the Configurations section of the ESP32-C6-DevKitC-1 board (`Documentation/platforms/risc-v/esp32c6/boards/esp32c6-devkitc/index.rst#L80`) explaining how to use it?

We will test this PR for ESP32-H2 too and insert a test for them in our internal CI too.

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+  pcnt_ll_set_edge_action(priv->dev, priv->config->pcntid, 1, PCNT_CHANNEL_EDGE_ACTION_INCREASE, PCNT_CHANNEL_EDGE_ACTION_DECREASE);
+  pcnt_ll_set_level_action(priv->dev, priv->config->pcntid, 1, PCNT_CHANNEL_LEVEL_ACTION_KEEP, PCNT_CHANNEL_LEVEL_ACTION_INVERSE);
+
+  /* Configure GPIO pins as Input with Pull-Up enabled */
```
GPIO config and matrix settings can go in the initialization function. Those should be configured only once during startup.

Vajnar replied:
done

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+
+  /* Disable PCNT clock */
+
+  periph_module_disable(PERIPH_PCNT_MODULE);
```
Maybe disable events and interrupts? What do you think?

Vajnar replied:
yes, changed accordingly.

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+  position = priv->position;
+  count = (int16_t)(pcnt_ll_get_count(priv->dev, priv->config->pcntid) & 0xffff);
+
+  /* Return the position measurement */
```
```suggestion
  /* Update the position measurement */
```

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+
+static int esp_setup(struct qe_lowerhalf_s *lower)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
```
Missing DEBUGASSERT

Vajnar replied:
added

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+  int32_t position;
+  int16_t count;
+
+  DEBUGASSERT(lower && priv->inuse);
```
You can move DEBUGASSERT before setting `*priv`

Vajnar replied:
done

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+  uint32_t regval;
+
+  sninfo("Resetting position to zero\n");
+  DEBUGASSERT(lower && priv->inuse);
```
You can move DEBUGASSERT before setting *priv

Vajnar replied:
done

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+}
+
+/****************************************************************************
+ * Name: esp_reset
```
Missing docstring

Vajnar replied:
Could you, please, tell me, how it should look?

fdcavalcanti replied:
Sure. You are missing a description of the input parameters and returned value. See this example from `esp_mcpwm.c`

```
/****************************************************************************
 * Name: esp_motor_fault_set
 *
 * Description:
 *   Sets the fault state for the motor.
 *
 * Input Parameters:
 *   lower - Pointer to the motor channel lower-half data structure.
 *   fault - Fault value. Must be one of motor_fault_e enum.
 *
 * Returned Value:
 *   OK on success, otherwise a negated errno value is returned on
 *   any failure.
 *
 ****************************************************************************/
```

Vajnar replied:
fixed

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+
+/****************************************************************************
+ * Name: esp_position
+ *
```
Missing docstring

Vajnar replied:
fixed

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+ * Name: esp_qeinitialize
+ *
+ * Description:
+ *   Initialize a quadrature encoder interface.  This function must be
```
```suggestion
 *   Initialize a quadrature encoder interface. This function must be
```

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+
+  /* Make sure that the PCNT is in the shutdown state */
+
+  esp_shutdown((struct qe_lowerhalf_s *)priv);
```
I recommend checking if the peripheral module is disabled by default during startup. If it is, just set the `position`to 0 and that's it.

Vajnar replied:
I now enable clock and reset the PCNT peripheral during initialization instead of during setup as it affects all units, so I could have inadvertently reset the state of other units. Please, check if you agree.

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+
+  const struct esp_qeconfig_s *config; /* static configuration */
+
+  bool inuse; /* True: The lower-half driver is in-use */
```
Noticed you set `inuse = true` only once and that's it. Maybe name it to `initialized` instead?
I would say that `inuse` would be appropriate to set as true when you open the device, and then as false when closed. But only if there's actual need for it.

Vajnar replied:
Yes, it should be set when opening the device to protect against double opening of the same unit. This would be problematic as during opening the position is zeroed. Please, check if you agree with current implementation.

fdcavalcanti replied:
Looks good!

Vajnar commented:
Hi @tmedicci,
> I'm not sure how you tested it, please describe it in detail. Particularly, I would like to know which testing application you used (if public, of course) and how we can reproduce and validate the testing from our side.

My current test setup features _ESP32-C6-DevKitC-1_ together with [Toshiba TB6612FNG](https://www.sparkfun.com/products/14451) motor driver and [motor with encoder](https://www.sparkfun.com/products/16413). During development I also used 2 channel oscilloscope to verify exact number of pulses from the Quadrature Encoder during motor rotation for comparison with pulse counter value in PCNT.

I tested just single unit. I don't have test setup to verify working of all 4 units simultaneously.

> In this same direction, can you please add a `defconfig` file for it in `nuttx/boards/risc-v/esp32c6/esp32c6-devkitc/configs/` and add an entry in the Configurations section of the ESP32-C6-DevKitC-1 board (`Documentation/platforms/risc-v/esp32c6/boards/esp32c6-devkitc/index.rst#L80`) explaining how to use it?

Yes, I can add it. There is also an example application, should I add it too?

> We will test this PR for ESP32-H2 too and insert a test for them in our internal CI too.

OK

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,593 @@
+ *
+ ****************************************************************************/
+
+static struct esp_lowerhalf_s *esp_pcnt2lower(int pcnt)
```
Make sure to move this. Not actually an implementation issue, but makes navigating the source code easier.

Vajnar replied:
I moved the part to initialization.

fdcavalcanti commented:
Hi @Vajnar, implementation wise it looks good to me. Like Tiago said previously, please add the defconfig and docs as well, we'll come up with a test routine for this.

Vajnar commented:
Hi @fdcavalcanti,

> Hi @Vajnar, implementation wise it looks good to me. Like Tiago said previously, please add the defconfig and docs as well, we'll come up with a test routine for this. Check for errors on CI as well.

I've added the `defconfig` along with entry to documentation. I checked CI and all formatting issues should be fixed. Could You, please, re-run the checks to make sure?

There is one more thing. In the last commit I made it possible to initialize all units. Previously it was possible to only instantiate the first one. Please, could you take a quick look?

fdcavalcanti commented:
> Hi @fdcavalcanti,
> 
> > Hi @Vajnar, implementation wise it looks good to me. Like Tiago said previously, please add the defconfig and docs as well, we'll come up with a test routine for this. Check for errors on CI as well.
> 
> I've added the `defconfig` along with entry to documentation. I checked CI and all formatting issues should be fixed. Could You, please, re-run the checks to make sure?
> 
> There is one more thing. In the last commit ([4fbf672](https://github.com/apache/nuttx/commit/4fbf672963439f1f2cac57d93d963430e7e6c67e)), I made it possible to initialize all units. Previously it was possible to only instantiate the first one. Please, could you take a quick look?

Looks ok for me. During your test, did you use all 4 units?

fdcavalcanti commented:
> Hi @fdcavalcanti,
> 
> > > Hi @fdcavalcanti,
> > > > Hi @Vajnar, implementation wise it looks good to me. Like Tiago said previously, please add the defconfig and docs as well, we'll come up with a test routine for this. Check for errors on CI as well.
> > > 
> > > 
> > > I've added the `defconfig` along with entry to documentation. I checked CI and all formatting issues should be fixed. Could You, please, re-run the checks to make sure?
> > > There is one more thing. In the last commit ([4fbf672](https://github.com/apache/nuttx/commit/4fbf672963439f1f2cac57d93d963430e7e6c67e)), I made it possible to initialize all units. Previously it was possible to only instantiate the first one. Please, could you take a quick look?
> > 
> > 
> > Looks ok for me. Just change the logging in `esp_board_qencoder.c` from sninfo to syslog. During your test, did you use all 4 units?
> 
> The logging is changed. I also changed initialization in board bringup to unify it with other peripherals.
> 
> During testing I used all 4 units simultaneously.

The syslog comment was for the board source files, not the actual driver. It's okay to use sensors debug there.

fdcavalcanti commented:
@Vajnar working on tests. Will soon let you know of more necessary modifications and overall results.
So far, very well done!

Vajnar commented:
> @Vajnar working on tests. Will soon let you know of more necessary modifications and overall results. So far, very well done!

Thank you! I changed the logging as you requested. Hopefully I got it right this time.

fdcavalcanti commented on arch/risc-v/src/common/espressif/esp_qencoder.c:
```diff
@@ -0,0 +1,657 @@
+
+  /* Configure GPIO pins as Input with Pull-Up enabled */
+
+  esp_configgpio(priv->config->ch0_gpio, INPUT_FUNCTION_3 | PULLUP);
```
```suggestion
  esp_configgpio(priv->config->ch0_gpio, INPUT_FUNCTION_3 | PULLUP);
  esp_configgpio(priv->config->ch0_gpio, INPUT_FUNCTION | PULLUP);
```

Vajnar replied:
Done

fdcavalcanti commented:
> But I expect that the commits should be quashed into single one or some small final set of commits, i.e. one in generic part `arch/risc-v/src/common/espressif`, one in the common ESP32C6 code `boards/risc-v/esp32c6/common/` and the last one for actual devkit mapping `boards/risc-v/esp32c6/esp32c6-devkitc` if the changes can be introduced such way then sources builds even after individual patches. Then the final state should be compared to actual final state after all these incremental patches to check that nothing has been forgotten.

Squashing two two commits is good enough. One for the driver on arch and one for C6 board support.

Vajnar commented:
Hi @fdcavalcanti, @tmedicci 
> Squashing two two commits is good enough. One for the driver on arch and one for C6 board support.

Should I proceed with the squashing?

fdcavalcanti commented:
> Hi @fdcavalcanti, @tmedicci 
> > Squashing two two commits is good enough. One for the driver on arch and one for C6 board support.
> 
> Should I proceed with the squashing?

Yes

Vajnar commented:
> > Hi @fdcavalcanti, @tmedicci
> > > Squashing two two commits is good enough. One for the driver on arch and one for C6 board support.
> > 
> > 
> > Should I proceed with the squashing?
> 
> Yes

Done
